File: src/kernel/sim-clock.ts
import type { EventBus } from './event-bus';

const MIN_RATE = 0.001;
const MAX_RATE = 100;
const INITIAL_TIME = -4500;

export interface SimClockConfig {
  /** Maximum real-time budget per advance call in seconds (adaptive tick rate).
   *  Caps dtReal to avoid runaway simulation when at max speed.
   *  Defaults to Infinity (no cap). */
  maxFrameBudget?: number;
}

export class SimClock {
  t: number = INITIAL_TIME;
  rate: number = 1;
  paused: boolean = false;

  /** Adaptive tick rate budget cap (seconds). */
  private readonly maxFrameBudget: number;

  /** Last effective real dt (after cap). */
  private _lastDtReal: number = 0;

  private bus: EventBus;

  constructor(bus: EventBus, config?: SimClockConfig) {
    this.bus = bus;
    this.maxFrameBudget = config?.maxFrameBudget ?? Infinity;
  }

  advance(dtReal: number): void {
    if (this.paused) return;

    // Adaptive tick rate: cap dt to avoid runaway when at max speed
    const cappedDt = Math.min(dtReal, this.maxFrameBudget);
    this._lastDtReal = cappedDt;

    const dtMa = cappedDt * this.rate;
    this.t += dtMa;
    this.bus.emit('TICK', { timeMa: this.t, deltaMa: dtMa });
  }

  seekTo(t: number): void {
    this.t = t;
  }

  pause(): void {
    this.paused = true;
  }

  resume(): void {
    this.paused = false;
  }

  togglePause(): void {
    this.paused = !this.paused;
  }

  setRate(rate: number): void {
    this.rate = Math.max(MIN_RATE, Math.min(MAX_RATE, rate));
  }

  /** Get the last real dt (capped by adaptive tick rate). */
  get lastDtReal(): number {
    return this._lastDtReal;
  }
}

File: backend/GeoTime.Core/Kernel/SimClock.cs
namespace GeoTime.Core.Kernel;

/// <summary>
/// Simulation clock managing geological time (Ma).
/// Starts at -4500 Ma (Earth's formation).
/// </summary>
public sealed class SimClock(EventBus bus, double maxFrameBudget = double.PositiveInfinity)
{
    private const double MIN_RATE = 0.001;
    private const double MAX_RATE = 100;
    public const double INITIAL_TIME = -4500;

    public double T { get; set; } = INITIAL_TIME;
    public double Rate { get; set; } = 1;
    public bool Paused { get; set; }
    public double LastDtReal { get; private set; }

    public void Advance(double dtReal)
    {
        if (Paused) return;
        var capped = Math.Min(dtReal, maxFrameBudget);
        LastDtReal = capped;
        var dtMa = capped * Rate;
        T += dtMa;
        bus.Emit("TICK", new { timeMa = T, deltaMa = dtMa });
    }

    public void SeekTo(double t) => T = t;
    public void Pause() => Paused = true;
    public void Resume() => Paused = false;
    public void TogglePause() => Paused = !Paused;
    public void SetRate(double rate) => Rate = Math.Clamp(rate, MIN_RATE, MAX_RATE);
}
